### ISSUE TITLE
Error in Publish after JSON unmarshal error in subscribed message

### ISSUE THREAD

**Issue by markuslindenberg** [nats-io/nats#78]
When using EncodedConn with JSON encoding, a decoding error will set `Conn.err`: https://github.com/nats-io/nats/blob/ef5165913ac88c53c2e2be39f73f8314490ea5dd/enc.go#L196

After `err` is set, publishing messages will not succeed because `publish()` will abort if a global error exists:
https://github.com/nats-io/nats/blob/ef5165913ac88c53c2e2be39f73f8314490ea5dd/nats.go#L1216

Use this test code to reproduce:
```go
package main

import (
	"fmt"
	"github.com/nats-io/nats"
	"time"
)

func main() {
	nc, _ := nats.Connect("nats://localhost:4222")
	c, _ := nats.NewEncodedConn(nc, "json")

	c.Subscribe("test", func(msg interface{}) {
		fmt.Println("Received", msg)
	})
	time.Sleep(500 * time.Millisecond)

	toPublish := [][]byte{
		[]byte(`{"msg": 1}`),
		[]byte(`{"msg": 2}`),
		[]byte(`foo`),
		[]byte(`{"msg": 4}`),
		[]byte(`{"msg": 5}`),
	}

	for i, data := range toPublish {
		err := nc.PublishMsg(&nats.Msg{
			Subject: "test",
			Data:    data,
		})
		fmt.Printf("Published #%d %s, error: %v\n", i, data, err)
		time.Sleep(500 * time.Millisecond)

	}

}
```

Also, I wonder how exactly the error handling is meant to work. If i do a `Publish()` (or any other method), I'd expect to only receive errors regarding the current operation, not any global errors. On the other hand, when the connection to gnatsd goes away, `Publish()` while the client ist disconnected will succeed without returning an error although the message was not sent to gnatsd.

**Comment by derekcollison** [comment]
Messages are buffered for sometime when disconnected and will be sent on a reconnect. I am looking into the global err vs local handling.

**Comment by markuslindenberg** [comment]
Great, thanks.

I didn't know about the buffering, love that the client is trying to do the right thing by itself. :)

**Comment by AnatoliiPrylutskyi** [comment]
Hi Derek!

I bumped on the same error in this line: 
https://github.com/nats-io/nats/blob/ef5165913ac88c53c2e2be39f73f8314490ea5dd/enc.go#L82

It was confusing to me debugging this issue in application. I think removing of global errors from subscribe and publishing is good idea. Encode/Decode error shoudn't affect next message.

I can create PR for this fix.

Thanks

**Comment by derekcollison** [comment]
I agree, on my list. Deep in some changes that are staged in a branch around core functionality for the Go client, but will move to this issue very soon.

**Comment by kozlovic** [comment]
I think that this may have been fixed in #131 and more generally in #158. Can I close this issue?

**Comment by mcqueary** [comment]
Closing per @kozlovic above comments that this is addressed in #131 and #158. @AnatoliiPrylutskyi or @markuslindenberg, please comment if you find this is not true and we can re-open.
